python/cuml/cluster/dbscan_mg.pyx

```diff
@@ -0,0 +1,204 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# distutils: language = c++
+
+from cuml.cluster import DBSCAN
+
+import ctypes
+import numpy as np
+import cupy as cp
+
+from libc.stdint cimport uintptr_t, int64_t
+
+from cuml.common.array import CumlArray
+from cuml.common.doc_utils import generate_docstring
+from cuml.raft.common.handle cimport handle_t
+from cuml.common import input_to_cuml_array
+from cuml.common import using_output_type
+
+
+cdef extern from "cuml/cluster/dbscan_mg.hpp" \
+        namespace "ML::Dbscan::opg":
+
+    cdef void fit(handle_t& handle,
+                  float *input,
+                  int n_rows,
+                  int n_cols,
+                  float eps,
+                  int min_pts,
+                  int *labels,
+                  int *core_sample_indices,
+                  size_t max_mbytes_per_batch,
+                  int verbosity) except +
+
+    cdef void fit(handle_t& handle,
+                  double *input,
+                  int n_rows,
+                  int n_cols,
+                  double eps,
+                  int min_pts,
+                  int *labels,
+                  int *core_sample_indices,
+                  size_t max_mbytes_per_batch,
+                  int verbosity) except +
+
+    cdef void fit(handle_t& handle,
+                  float *input,
+                  int64_t n_rows,
+                  int64_t n_cols,
+                  double eps,
+                  int min_pts,
+                  int64_t *labels,
+                  int64_t *core_sample_indices,
+                  size_t max_mbytes_per_batch,
+                  int verbosity) except +
+
+    cdef void fit(handle_t& handle,
+                  double *input,
+                  int64_t n_rows,
+                  int64_t n_cols,
+                  double eps,
+                  int min_pts,
+                  int64_t *labels,
+                  int64_t *core_sample_indices,
+                  size_t max_mbytes_per_batch,
+                  int verbosity) except +
+
+
+class DBSCANMG(DBSCAN):
+    """
+    A Multi-Node Multi-GPU implementation of DBSCAN
+
+    NOTE: This implementation of DBSCAN is meant to be used with an
+    initialized cumlCommunicator instance inside an existing distributed
+    system. Refer to the Dask BSCAN implementation in
+    `cuml.dask.cluster.dbscan`.
+    """
+
+    def __init__(self, **kwargs):
+        super(DBSCANMG, self).__init__(**kwargs)
+
+    @generate_docstring(skip_parameters_heading=True)
+    def fit(self, X, out_dtype="int32"):
+        """
+        Perform DBSCAN clustering in a multi-node multi-GPU setting.
+
+        Parameters
+        ----------
+        out_dtype: dtype Determines the precision of the output labels array.
+            default: "int32". Valid values are { "int32", np.int32,
+            "int64", np.int64}.
+        """
+        self._set_base_attributes(n_features=X)
```

Comment:
I think what you've done sounds great. Though, rather than exposing the OPG argument on the public API, why don't we move `fit()` into a private/internal function by renaming it to `_fit()` and have the fit()` function in `dbscan.pyx` just proxy down to `_fit()` with `opg=False`. Then in the dask wrapper, you can just have each worker call `_fit(..., opg=True)`. Another alternative would be to keep a `dbscan_mg.pyx` file to be consistent with other estimators but just override the `fit()` funciton in the baseclass to call `_fit(..., opg=True)`. I think I might prefer the latter idea just because it keeps the consistent design but removes all the duplication.

Reply:
Ok, I did the latter.